### PR TITLE
Refactor TLA support

### LINKED ISSUE #339: Redo TLA support by taking it from bellard/quickjs
Turns out ours is a bit crude. `JS_EvalFunction` won't return a promise in all cases, for example.

More importantly, it would bring the forks closer toghether.

Relevant commits:

https://github.com/bellard/quickjs/commit/6e4931c4ad6c1f06cd1e46ecbf98e04cbb8b4ba7
https://github.com/bellard/quickjs/commit/00967aac2408af0d3e591918b25e700229c9cdb3
https://github.com/bellard/quickjs/commit/67723c93e4241508ad1ff9bfad142e39f3f2cf88
https://github.com/bellard/quickjs/commit/84058766e9f247cac7b00543feeb10f3bba8120a
https://github.com/bellard/quickjs/commit/e44b793e3817766d9045ed2a776658fcbe0f2790

### PR DESCRIPTION
**Pull request by QuiiBz**
Closes https://github.com/quickjs-ng/quickjs/issues/339

392bbff5470bc9a9fad81aa685c931353272752e essentially reverts https://github.com/quickjs-ng/quickjs/pull/6 and the next commits apply progressively the commits mentioned in https://github.com/quickjs-ng/quickjs/issues/339.

### CODE REVIEW

**Comment by saghul** [comment]
Thank you for doing this! ❤️

**Comment by QuiiBz** [comment]
I'm planning to use QuickJS/QuickJS-NG soon, so contributing to this project makes sense!

The only thing not implemented is TLA support in the REPL. The code from this fork seems quite different at a first glance, but I'll try to make it work soon. Not sure if that's a blocker for this PR.

**Comment by saghul** [comment]
I think not but I'll take a closer look. Thanks again!

**Comment by saghul** [comment]
The only 2 missing ones are: 

https://github.com/bellard/quickjs/commit/e44b793e3817766d9045ed2a776658fcbe0f2790
https://github.com/bellard/quickjs/commit/00967aac2408af0d3e591918b25e700229c9cdb3

I think we should at least include https://github.com/bellard/quickjs/commit/00967aac2408af0d3e591918b25e700229c9cdb3 here and the bits in https://github.com/bellard/quickjs/commit/e44b793e3817766d9045ed2a776658fcbe0f2790 that add the new eval flag. The REPL changes can be a subsequent PR.

**Comment by QuiiBz** [comment]
Just added `os.sleepAsync`, the other changes in those two commits were already there (except the REPL changes).

**Review by saghul — APPROVED** [review]
LGTM. @chqrlie can you PTAL too?

**Comment by chqrlie** [comment]
> LGTM. @chqrlie can you PTAL too?

Looking at this... Takes some time to ensure no unseen regressions, should be done soon.

File: run-test262.c [hunk 1]
```diff
@@ -1207,50 +1207,46 @@ static int eval_buf(JSContext *ctx, const char *buf, size_t buf_len,
...
-                JSPromiseStateEnum s = JS_PromiseState(ctx, promise);
-                if (s == JS_PROMISE_FULFILLED)
-                    res_val = JS_UNDEFINED;
-                else if (s == JS_PROMISE_REJECTED)
-                    res_val = JS_Throw(ctx, JS_PromiseResult(ctx, promise));
-                else
-                    res_val = JS_EXCEPTION;
-                break;
-            }
-        }
```
Comment (chqrlie) [inline run-test262.c:1253]: There is patch missing in function `run_test262_harness_test`: the lines
```
    } else {
        JS_FreeValue(ctx, res_val);
        for(;;) {
            JSContext *ctx1;
            ret = JS_ExecutePendingJob(JS_GetRuntime(ctx), &ctx1);
            if (ret < 0) {
          js_std_dump_error(ctx1);
          ret_code = 1;
            } else if (ret == 0) {
          break;
            }
        }
    }
```
Should be changed to
```
    } else {
        JSValue promise = JS_UNDEFINED;
        if (is_module) {
            promise = res_val;
        } else {
            JS_FreeValue(ctx, res_val);
        }
        for(;;) {
            JSContext *ctx1;
            ret = JS_ExecutePendingJob(JS_GetRuntime(ctx), &ctx1);
            if (ret < 0) {
                js_std_dump_error(ctx1);
                ret_code = 1;
            } else if (ret == 0) {
                break;
            }
        }
        /* dump the error if the module returned an error. */
        if (is_module) {
            JSPromiseStateEnum state = JS_PromiseState(ctx, promise);
            if (state == JS_PROMISE_REJECTED) {
                JS_Throw(ctx, JS_PromiseResult(ctx, promise));
                js_std_dump_error(ctx);
                ret_code = 1;
            }
        }
        JS_FreeValue(ctx, promise);
    }
```
I am not sure this is necessary, but Fabrice's version has these changes. Can you verify?

Comment (saghul) [reply run-test262.c:1253]: It sure looks necessary, good catch!

Comment (QuiiBz) [reply run-test262.c:1253]: Just pushed the [patch](https://github.com/bellard/quickjs/commit/3c2cfabfc74e8af1f21db93884ed1ad9b6388a8c), indeed good catch!

**Comment by saghul** [comment]
Alright, I rebased this and added the missing REPL bits. @chqrlie PTAL!

Something I noticed is the difference in return values now:

```
qjs > os.readdir('.')
{
  value: [
    [
      '.',                   '..',                  'libregexp.c',
      'v8.sh',               'repl.js',             'test262.conf',
      'getopt_compat.h',     'quickjs.c',           'CMakeLists.txt',
      'LICENSE',             'run-test262.c',       'Makefile',
      'v8.txt',              'libunicode-table.h',  'qjs.c',
      'quickjs-c-atomics.h', 'libunicode.c',        'tests',
      'libbf.h',             'dirent_compat.h',     'cutils.c',
      'quickjs-libc.c',      'qjsc.c',              'gen',
      '.gitmodules',         'v8-tweak.js',         'README.md',
      'libregexp-opcode.h',  'list.h',              'a.out',
      '.gitignore',          'quickjs-opcode.h',    'libregexp.h',
      'examples',            't.js',                '.github',
      'quickjs-atom.h',      'doc',                 'quickjs.h',
      'unicode_gen_def.h',   'libunicode.h',        'v8.js',
      'test262_errors.txt',  'test262-fast.conf',   'build',
      'test262',             '.git',                'quickjs-libc.h',
      't.c',                 'unicode_download.sh', 'cutils.h',
      'unicode_gen.c',       'libbf.c'
    ],
    0
  ]
}
qjs >
```

vs before:

```
qjs > os.readdir('.')
[
  [
    '.',                   '..',                  'libregexp.c',
    'v8.sh',               'repl.js',             'test262.conf',
    'getopt_compat.h',     'quickjs.c',           'CMakeLists.txt',
    'LICENSE',             'run-test262.c',       'Makefile',
    'v8.txt',              'libunicode-table.h',  'qjs.c',
    'quickjs-c-atomics.h', 'libunicode.c',        'tests',
    'libbf.h',             'dirent_compat.h',     'cutils.c',
    'quickjs-libc.c',      'qjsc.c',              'gen',
    '.gitmodules',         'v8-tweak.js',         'README.md',
    'libregexp-opcode.h',  'list.h',              'a.out',
    '.gitignore',          'quickjs-opcode.h',    'libregexp.h',
    'examples',            't.js',                '.github',
    'quickjs-atom.h',      'doc',                 'quickjs.h',
    'unicode_gen_def.h',   'libunicode.h',        'v8.js',
    'test262_errors.txt',  'test262-fast.conf',   'build',
    'test262',             '.git',                'quickjs-libc.h',
    't.c',                 'unicode_download.sh', 'cutils.h',
    'unicode_gen.c',       'libbf.c'
  ],
  0
]
qjs >
```

Upstream is correct so it seems something is amiss here.

**Comment by saghul** [comment]
Simpler case:

```
qjs > var x;
{ value: undefined }
```

**Comment by saghul** [comment]
Ah https://github.com/bellard/quickjs/commit/00967aac2408af0d3e591918b25e700229c9cdb3

**Comment by saghul** [comment]
Alright, this is now ready to do AFAICT! @chqrlie can you PTAL? I'd like to cut a release after this goes in.

**Comment by chqrlie** [comment]
> Alright, this is now ready to do AFAICT! @chqrlie can you PTAL? I'd like to cut a release after this goes in.

Very busy at the moment, will try to review by Friday night.

**Comment by saghul** [comment]
Gentle ping @chqrlie :-)

**Comment by saghul** [comment]
So, our implementation matches upstream now, 262 passes and I also tested with txiki.js: https://github.com/saghul/txiki.js/pull/540

I'm going to merge this and if we need to make tweaks to it we can dot hat as we do any other bugfix.

**Comment by chqrlie** [comment]
> So, our implementation matches upstream now, 262 passes and I also tested with txiki.js: [saghul/txiki.js#540](https://github.com/saghul/txiki.js/pull/540)
> 
> I'm going to merge this and if we need to make tweaks to it we can dot hat as we do any other bugfix.

OK for me, I was about to write the same thing :)

**Pull request merged**